presto-main/src/main/java/com/facebook/presto/resourcemanager/DistributedQueryInfoResource.java

```diff
@@ -0,0 +1,152 @@
+import com.google.common.util.concurrent.ListeningExecutorService;
+
+import javax.annotation.security.RolesAllowed;
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Future;
+
+import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static com.facebook.presto.server.security.RoleType.ADMIN;
+import static com.facebook.presto.server.security.RoleType.USER;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.net.HttpHeaders.X_FORWARDED_PROTO;
+import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.Response.Status.NOT_FOUND;
+
+@Path("/v1/queryState")
+@RolesAllowed({USER, ADMIN})
+public class DistributedQueryInfoResource
+{
+    private static final Logger log = Logger.get(DistributedQueryInfoResource.class);
+    private static final JsonCodec<List<QueryStateInfo>> JSON_CODEC = JsonCodec.listJsonCodec(QueryStateInfo.class);
+    private final ResourceManagerClusterStateProvider clusterStateProvider;
+    private final InternalNodeManager internalNodeManager;
+    private ListeningExecutorService executor;
+    private final ResourceManagerProxy proxyHelper;
+
+    @Inject
+    public DistributedQueryInfoResource(ResourceManagerClusterStateProvider clusterStateProvider, InternalNodeManager internalNodeManager,
+            @ForResourceManager ListeningExecutorService executor, ResourceManagerProxy proxyHelper)
+    {
+        this.clusterStateProvider = requireNonNull(clusterStateProvider, "clusterStateProvider is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.proxyHelper = requireNonNull(proxyHelper, "proxyHelper is null");
+    }
+
+    @GET
+    public void getAllQueryInfo(@QueryParam("user") String user,
+            @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
+            @Context HttpServletRequest servletRequest,
+            @Context UriInfo uriInfo,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        try {
+            Set<InternalNode> coordinators = internalNodeManager.getCoordinators();
+            List<ListenableFuture<List<QueryStateInfo>>> queryStateInfoFutureList = new ArrayList<>();
```

Comment:
In general we use immutable data structures.  Here I would either stream through the coordinators, perform the action and collect() it into an `toImmutableList()`, or use an ImmutableList.Builder

Reply:
I have to get used to immutable list :) I made the changes to use ImmutableList.Builder